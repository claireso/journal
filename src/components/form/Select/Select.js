import React from 'react'
import PropTypes from 'prop-types'

import { IconAngleRight } from '@components/Icons'

import { GroupInline } from '../Group'
import Label from '../Label'

import * as S from './Select.styles'

const Select = (props = {}) => {
  return (
    <GroupInline>
      <Label htmlFor={props.name}>{props.label}</Label>
      <S.SelectWrapper>
        <S.StyledSelect id={props.name} name={props.name} defaultValue={props.value} required>
          {props.options.map((option, index) => {
            return (
              <option key={index} value={option.value}>
                {option.label}
              </option>
            )
          })}
        </S.StyledSelect>
        <IconAngleRight />
      </S.SelectWrapper>
    </GroupInline>
  )
}

Select.propTypes = {
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired
    })
  ).isRequired,
  value: PropTypes.string
}

export default Select
