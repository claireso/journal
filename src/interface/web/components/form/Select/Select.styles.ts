import { styled } from '@web/theme'

export const SelectWrapper = styled('div', {
  position: 'relative',
  svg: {
    position: 'absolute',
    top: '50%',
    right: '0.9rem',
    bottom: 0,
    transform: 'translate(0, -50%) rotate(90deg)',
    pointerEvents: 'none'
  }
})

export const StyledSelect = styled('select', {
  appearance: 'none',
  background: '$white',
  border: '1px solid $gray200',
  borderRadius: '0.5rem',
  boxShadow: '$2',
  cursor: 'pointer',
  fontFamily: '$sansSerif',
  fontSize: '$4',
  outline: 'none',
  padding: '1.1rem 2.6rem 1.1rem 1.1rem',
  transition: '$borderColor',
  '&:focus': {
    borderColor: '$gray300'
  }
})
