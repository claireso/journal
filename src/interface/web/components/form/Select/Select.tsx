import React from 'react'

import Icon from '@web/components/Icons'

import Group from '../Group'
import Label from '../Label'

import * as cls from './styles.css'

interface Option {
  label: string
  value: string
}

interface SelectProps {
  name: string
  label: string
  options: Option[]
  value?: string
}

const Select = ({ name, label, value, options }: SelectProps) => {
  return (
    <Group inline>
      <Label htmlFor={name}>{label}</Label>
      <div className={cls.wrapper}>
        <select className={cls.select} id={name} name={name} defaultValue={value} required>
          {options.map((option, index) => {
            return (
              <option key={index} value={option.value}>
                {option.label}
              </option>
            )
          })}
        </select>
        <div className={cls.icon}>
          <Icon name="arrow-down" size="sm" />
        </div>
      </div>
    </Group>
  )
}

export default Select
