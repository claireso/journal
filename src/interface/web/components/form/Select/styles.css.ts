import { cmpStyles, tokens, radius, p, pr } from '@web/theme'

export const wrapper = cmpStyles({
  position: 'relative'
})

export const select = cmpStyles({
  appearance: 'none',
  cursor: 'pointer',
  borderWidth: '1px',
  borderStyle: 'solid',
  transition: 'border 150ms ease-out',
  outline: 'none',
  fontSize: tokens.fonts.size.base,
  fontFamily: tokens.fonts.family.sansSerif,
  background: tokens.colors.white,
  boxShadow: tokens.shadows.input,
  borderColor: tokens.colors.neutral.extralight,
  ...radius('size-1'),
  ...p('size-3'),
  ...pr('size-7'),
  selectors: {
    '&:focus': {
      borderColor: tokens.colors.neutral.light
    }
  }
})

export const icon = cmpStyles({
  position: 'absolute',
  top: '50%',
  right: tokens.sizes['size-2'],
  bottom: 0,
  height: tokens.sizes['size-4'],
  transform: 'translate(0, -50%)',
  pointerEvents: 'none'
})
