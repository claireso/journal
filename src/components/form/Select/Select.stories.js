import React from 'react'

import Select from './index'

export default {
  title: 'Form/Select'
}

const options = [
  {
    label: 'Value 1',
    value: '1'
  },
  {
    label: 'Value 2',
    value: '2'
  }
]

export const Basic = () => <Select label="Label" name="position" options={options} />
