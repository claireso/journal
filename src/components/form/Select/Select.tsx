import React from 'react'

import { IconAngleRight } from '@components/Icons'

import { GroupInline } from '../Group'
import Label from '../Label'

import * as S from './Select.styles'

interface Option {
  label: string
  value: string
}

interface SelectProps {
  name: string
  label: string
  options: Option[]
  value?: string
}

const Select = ({ name, label, value, options }: SelectProps) => {
  return (
    <GroupInline>
      <Label htmlFor={name}>{label}</Label>
      <S.SelectWrapper>
        <S.StyledSelect id={name} name={name} defaultValue={value} required>
          {options.map((option, index) => {
            return (
              <option key={index} value={option.value}>
                {option.label}
              </option>
            )
          })}
        </S.StyledSelect>
        <IconAngleRight />
      </S.SelectWrapper>
    </GroupInline>
  )
}

export default Select
