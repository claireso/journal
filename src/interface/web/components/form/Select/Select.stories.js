import React from 'react'

import Select from './index'

const params = {
  title: 'Form/Select'
}

export default params

const options = [
  {
    label: 'Value 1',
    value: '1'
  },
  {
    label: 'Value 2',
    value: '2'
  }
]

export const Basic = () => <Select label="Label" name="position" options={options} />
