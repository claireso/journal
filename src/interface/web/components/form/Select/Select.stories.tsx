import type { Meta, StoryObj } from '@storybook/react-webpack5'

import Select from './Select'

const meta: Meta<typeof Select> = {
  title: 'Form/Select',
  component: Select
}

export default meta
type Story = StoryObj<typeof Select>

export const Basic: Story = {
  args: {
    label: 'Label',
    name: 'position',
    options: [
      {
        label: 'Value 1',
        value: '1'
      },
      {
        label: 'Value 2',
        value: '2'
      }
    ]
  }
}
