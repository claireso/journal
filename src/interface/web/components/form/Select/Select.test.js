import { render } from '@testing-library/react'

import Select from './index'

describe('<Select />', () => {
  it('should render select', () => {
    const props = {
      label: 'Position',
      name: 'position',
      options: [
        {
          label: 'left',
          value: 'left'
        },
        {
          label: 'right',
          value: 'right'
        }
      ],
      value: 'right'
    }

    const { asFragment } = render(<Select {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
