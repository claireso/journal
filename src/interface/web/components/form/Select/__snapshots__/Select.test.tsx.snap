// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render select 1`] = `
<DocumentFragment>
  <div
    class="styles__12zyxsm3 styles_layerStyle__12zyxsm0 styles_wrapper_inline_true__12zyxsm4 styles_layerStyle__12zyxsm1"
  >
    <label
      class="styles_layerStyle__17qmo0z0"
      for="position"
    >
      Position
    </label>
    <div
      class="styles_layerStyle__1v4xt870"
    >
      <select
        class="styles_layerStyle__1v4xt871"
        id="position"
        name="position"
        required=""
      >
        <option
          value="left"
        >
          left
        </option>
        <option
          selected=""
          value="right"
        >
          right
        </option>
      </select>
      <div
        class="styles_layerStyle__1v4xt872"
      >
        <svg
          aria-hidden="true"
          class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_sm__19o76jub styles_layerStyle__19o76ju2 styles_icon_variant_default__19o76juh styles_layerStyle__19o76ju8"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m19.5 8.25-7.5 7.5-7.5-7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
